### PR TITLE
Fix two blocking issues

### LINKED ISSUE #17769: Only 1 soldier can pass the service depot tiles at a time
To reproduce get a 5 stack of infantry next to a service depot, the order them to move to the opposite side of it

### PR DESCRIPTION
**Pull request by pchote**
The second commit fixes #17769.
The third and fourth commits fix https://github.com/OpenRA/OpenRA/pull/17618#issuecomment-578504081 and other issues where groups of infantry block each other while moving

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/World/Locomotor.cs [hunk 1]
```diff
@@ -336,14 +336,14 @@ public SubCell GetAvailableSubCell(Actor self, CPos cell, BlockedByActor check,
...
-		bool IsBlockedBy(Actor self, Actor otherActor, Actor ignoreActor, BlockedByActor check, CellFlag cellFlag)
+		bool IsBlockedBy(Actor actor, Actor otherActor, Actor ignoreActor, CPos cell, BlockedByActor check, CellFlag cellFlag)
 		{
 			if (otherActor == ignoreActor)
 				return false;
 
 			// If the check allows: We are not blocked by units that we can force to move out of the way.
 			if (check <= BlockedByActor.Immovable && cellFlag.HasCellFlag(CellFlag.HasMovableActor) &&
-				self.Owner.Stances[otherActor.Owner] == Stance.Ally)
+			    actor.Owner.Stances[otherActor.Owner] == Stance.Ally)
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/World/Locomotor.cs]: Convert the spaces to a tab please.

**Comment by pchote** [comment]
Fixed.

**Review by teinarss — APPROVED** [review]

**Comment by deleted-user-1** [comment]
I did some testing and added the logging below to see when when in `Locomotor.GetAvailableSubCell` the path for blocked units would be used. Although I might be misunderstanding the code and thus my logging could be wrong, I was surprised that even only with one infantry soldier moving from a free cell to another free cell, the output very very often showed that the engine considers the path to be blocked (`Infantry blocked! ...`). It looks like it is blocking itself.

Another thing I noticed is that when you send two 5-stacks of infantry on paths so that they collide with the other 5-stack, the moment they collide they wait for some seconds before re-pathing around each other (they don't give up). The logging output shows that they can't find any subcell to go. Lastly the `..., use default Subcell` case never appears in logging output.

``` diff

diff --git a/OpenRA.Mods.Common/Traits/World/ActorMap.cs b/OpenRA.Mods.Common/Traits/World/ActorMap.cs
index 92491d758d..7617de98f0 100644
--- a/OpenRA.Mods.Common/Traits/World/ActorMap.cs
+++ b/OpenRA.Mods.Common/Traits/World/ActorMap.cs
@@ -277,29 +277,54 @@ public bool HasFreeSubCell(CPos cell, bool checkTransient = true)
 		public SubCell FreeSubCell(CPos cell, SubCell preferredSubCell = SubCell.Any, bool checkTransient = true)
 		{
 			if (preferredSubCell != SubCell.Any && !AnyActorsAt(cell, preferredSubCell, checkTransient))
+			{
+				Console.WriteLine("Not blocked, preferred SubCell is {0}", preferredSubCell);
 				return preferredSubCell;
+			}
 
 			if (!AnyActorsAt(cell))
+			{
+				Console.WriteLine("Not blocked, use default Subcell {0}", map.Grid.DefaultSubCell);
 				return map.Grid.DefaultSubCell;
+			}
 
 			for (var i = (int)SubCell.First; i < map.Grid.SubCellOffsets.Length; i++)
+			{
 				if (i != (int)preferredSubCell && !AnyActorsAt(cell, (SubCell)i, checkTransient))
+				{
+					Console.WriteLine("Not blocked, found free subcell {0}", (SubCell)i);
 					return (SubCell)i;
-
+				}
+			}
+				
+			Console.WriteLine("Not blocked, no free subcell found!");
 			return SubCell.Invalid;
 		}
 
 		public SubCell FreeSubCell(CPos cell, SubCell preferredSubCell, Func<Actor, bool> checkIfBlocker)
 		{
 			if (preferredSubCell != SubCell.Any && !AnyActorsAt(cell, preferredSubCell, checkIfBlocker))
+			{
+				Console.WriteLine("Infantry blocked! Preferred SubCell is {0}", preferredSubCell);
 				return preferredSubCell;
+			}
 
 			if (!AnyActorsAt(cell))
+			{
+				Console.WriteLine("Infantry blocked! Use default Subcell {0}", map.Grid.DefaultSubCell);
 				return map.Grid.DefaultSubCell;
+			}
 
 			for (var i = (byte)SubCell.First; i < map.Grid.SubCellOffsets.Length; i++)
+			{
 				if (i != (byte)preferredSubCell && !AnyActorsAt(cell, (SubCell)i, checkIfBlocker))
+				{
+					Console.WriteLine("Infantry blocked! Found free subcell {0}", (SubCell)i);
 					return (SubCell)i;
+				}
+			}
+			
+			Console.WriteLine("Infantry blocked! No free subcell found!");
 			return SubCell.Invalid;
 		}
 

```

**Review by deleted-user-1 — APPROVED** [review]
Fixes for #17769 and https://github.com/OpenRA/OpenRA/pull/17618#issuecomment-578504081 work.

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/d79fdd0a37cf5602e9acfda0fff060dcbb2dee9d...1727457f38b51a29e33cbd310efbf7495bef4603)
https://github.com/OpenRA/OpenRA/commit/85412a6d98d9cb4886073038e29615e27bbfe500
https://github.com/OpenRA/OpenRA/commit/5489d6909c879c2a2759ad12ff9b84c820a6462a
https://github.com/OpenRA/OpenRA/commit/c9dcf305c9ef2f8975517c8868c29f5251a5706b
https://github.com/OpenRA/OpenRA/commit/5a8964aabebfc6a76eccedc15839d6975479d707

**Pull request merged**